controllers/argocd/role.go

```diff
@@ -156,6 +176,86 @@ func (r *ReconcileArgoCD) reconcileRole(name string, policyRules []v1.PolicyRule
 	return roles, nil
 }
 
+func (r *ReconcileArgoCD) reconcileRoleForSupportedNamespaces(name string, policyRules []v1.PolicyRule, cr *argoprojv1a1.ArgoCD) ([]*v1.Role, error) {
+	var roles []*v1.Role
+
+	managedClusterArgoCDNamespaces, err := r.getManagedClusterArgoCDNamespaces(cr)
+
+	// create policy rules for each supported namespace for ArgoCD Server
+	for _, sourceNamespace := range cr.Spec.Server.SourceNamespaces {
+
+		namespace := &corev1.Namespace{}
+		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: sourceNamespace}, namespace); err != nil {
+			return nil, err
+		}
+
+		managedNamespace := false
+		// do not reconcile roles for namespaces already containing managed-by label
+		// as it already contains roles reconciled during reconcilation of ManagedNamespaces
+		for _, ns := range r.ManagedNamespaces.Items {
+			if reflect.DeepEqual(ns, namespace) {
+				managedNamespace = true
+				break
+			}
+		}
+		if managedNamespace {
+			continue
+		}
+
+		// do not reconcile roles for namespaces already containing managed-by-cluster-argocd label
+		// as it already contains roles reconciled during reconcilation of ManagedNamespaces
+		if err != nil {
+			log.Info("Could not find list of namespaces managed by ArgoCD as sourceNamespaces")
+			continue
+		}
+		for _, ns := range managedClusterArgoCDNamespaces.Items {
+			if reflect.DeepEqual(ns, namespace) {
+				managedNamespace = true
+				break
+			}
+		}
+		if managedNamespace {
+			continue
+		}
+		log.Info(fmt.Sprintf("Reconciling role for %s", namespace.Name))
+
+		// Update namespace with managed-by-cluster-argocd label
+		namespace.Labels[common.ArgoCDManagedByClusterArgoCDLabel] = cr.Name
```

Comment:
We are setting `cr.Name` as the value for the label. Added a new check for checking if another `cr.Name` is assigned to the label for avoiding overwrite issues.

Reply:
should we use the namespace instead of name? might be easier since namespace names are unique but you can have multiple instances with the same name in different namespaces

It would also be consistent with how we use the instance namespace for managed-by label

Reply:
I kept it the instance name as we could have multiple argocd instances in the same namespace as well. That way it would be difficult to understand which argocd instance is managing the namespace.

Reply:
I don't think you can have multiple argocd instances in one namespace 
that sounds like it would cause a lot of problems

Reply:
Makes sense, have updated it.